### DISCUSSION TITLE
All HTMLAttrDef and HTMLAttrXRef macros are removed mdn/content

### DISCUSSION THREAD

**Discussion by schalkneethling** [mdn/mdn-community#347] [Announcements]
From @OnkarRuikar

We've removed all `HTMLAttrDef` and `HTMLAttrXRef` macros from mdn/content. They've been converted to either markdown links or anchor tags. They are being marked deprecated in yari.

## Henceforth use markdown style links or anchor tags:

a. For `{{htmlattrxref("lang")}}` use `[lang](/en-US/docs/Web/HTML/Global_attributes#lang)` in Markdown and
`<a href="/en-US/docs/Web/HTML/Global_attributes#lang)"><code>lang</code></a>` in HTML tables.
b. For `{{htmlattrxref("srcset", "img")}}` use `[srcset](/en-US/docs/Web/HTML/Element/img#srcset)` in Markdown `<a href="/en-US/docs/Web/HTML/Element/img#srcset)"><code>srcset</code></a>` in HTML tables.

The conversion was not as straightforward as I thought it would be. Thanks to @teoli2003, we solved the issues quickly.

### Here is what we did:

1. Converted `{{HTMLAttrDef("abc")}}` to `abc`
2. Removed `attr-` from URL fragments `/en-US/docs/Web/HTML/Element/tag#attr-abc`
3. Converted HTMLAttrXRef macros:

a. Converted 1 argument instances to `/en-US/docs/Web/HTML/Global_attributes#lang`. Later we decided to convert these to `/en-US/docs/Web/HTML/Global_attributes/lang` (instead of a fragment, use the direct link)
b. Converted 3 and 4 argument instances. Because there was not a lot of these, I did it manually.
c. Converted 2 arg instances from `{{HTMLAttrXref("attr", "tag")}}` to `[attr](/en-US/docs/HTML/Element/tag#attr)`.

1. We used `[abc](#abc)`
2. For same-page links, we had to use anchor tags for the instances in HTML tables: `<a href="/en-US/../tag#attr)"><code>attr</code><a>`

> **Note:** Some macro instances were producing non-existing URLs because, over time, people put anything in the macro calls. Use PR assistants flaws section to find the broken URLs related to the task.

I used the following RegExes to semi-automate the task.

a. Convert 1 param macros:
Search: `\{\{ ?htmlattrxref ?\( ?["']([^,]*?)["']\)\}\}`
Replacement: `[$1](/en-US/docs/Web/HTML/Global_attributes#$1)`

b. Convert 2 param macros:
Search: `\{\{ ?htmlattrxref ?\( ?["']([^,\n}]*?)["'], ?["']([^,\n}]*?)["']\)\}\}`\
Replacement: `[$1](/en-US/docs/Web/HTML/Element/$2#$1)`

Then to update markdown style links in HTML tables:
Search: `(?<=<td>)(([\s\S](?!/td))*?)\[(.*)\]\((.*)\)`\
Replacement: `$1<a href="$4"><code>$3</code></a>`

**Comment by teoli2003** [comment]
A small nit:

Converted {{HTMLAttrDef("abc")}} to \`abc`

**Reply by schalkneethling** [reply]
Thanks, @teoli2003 - updated.

**Comment by yin1999** [comment]
I found there are some cases that the macro are wrapped with inline-code. We may need to use regexp like ``\`?\{\{ ?htmlattrxref ?\( ?["']([^,]*?)["']\)\}\}\`?`` to avoid repeat wrapping.

**Reply by OnkarRuikar** [reply]
Also some old PRs that are getting merged now are re-introducing the macro instances to `mdn/content`.
